ports/raspberrypi/common-hal/digitalio/DigitalInOut.c

```diff
@@ -42,8 +42,21 @@
 #include "bindings/cyw43/__init__.h"
 #define IS_CYW(self) ((self)->pin->base.type == &cyw43_pin_type)
 
-const mcu_pin_obj_t *common_hal_digitalio_validate_pin(mp_obj_t obj) {
-    return validate_obj_is_free_pin_including_cyw43(obj, MP_QSTR_pin);
+const mcu_pin_obj_t *common_hal_digitalio_validate_pin(mp_obj_t obj, qstr arg_name) {
+    return validate_obj_is_free_pin_including_cyw43(obj, arg_name);
+}
+#endif
+
+#if CIRCUITPY_TCA9555R
+#include "bindings/tca9555r/__init__.h"
+#define IS_TCA(self) ((self)->pin->base.type == &tca_pin_type)
+
+const mcu_pin_obj_t *common_hal_digitalio_validate_pin(mp_obj_t obj, qstr arg_name) {
+    return validate_obj_is_free_pin_including_tca(obj, arg_name);
+}
+
+const mcu_pin_obj_t *common_hal_digitalio_validate_pin_or_none(mp_obj_t obj, qstr arg_name) {
+    return validate_obj_is_free_pin_including_tca_or_none(obj, arg_name);
```

Comment:
Here's an example of how introducing a "protocol" will help. Instead of having a separate preprocessor ifdef for each possible kind of IO expander, it would become a call like `return mp_proto_get(obj, MP_QSTR_Pin) != NULL`. This protocol would be implemented by microcontroller.Pin, as well as the cyw43 pin, the tca9555 pin, and the PCA9554.

The protocol structure would have a number of function pointers, for getting & setting value, direction, & pull.

Then the common_hal_digitalio_digitalinout functions call via the function pointer instead of inlining the code for TCA, microcontroller.Pin, etc.

If this explanation doesn't make sense on its own, you can get more of an idea by looking at e.g., how displayio framebuffers use protocols. I'm also happy to expand on the concept more.

Reply:
Yes, I can see how changing to your proposed protocol here would be much better

Reply:
Would these IO expander implementations be made available to user code? If they have a Python dictionary, then you could do a lookup in the dictionary instead of having a new protocol. That's how you'd do it if you were using a Python object.